aws_lambda_powertools/utilities/feature_toggles/configuration_store.py

```diff
@@ -0,0 +1,207 @@
+            cache_seconds (int): cache expiration time, how often to call AppConfig to fetch latest configuration
+            region_name (str): aws region where the configuration resides in
+        """
+        self._cache_seconds = cache_seconds
+        self._conf_name = conf_name
+        config = None
+        if region_name is not None:
+            config = Config(region_name=region_name)
+        self._conf_store = AppConfigProvider(environment=environment, application=service, config=config)
+
+    def _validate_json_schema(self, schema: str) -> bool:
+        ## todo
+        return True
+
+    def _match_by_action(self, action: str, restriction_value: Any, context_value: Any) -> bool:
+        mapping_by_action = {
+            ACTION.EQUALS.value: lambda a, b: a == b,
+            ACTION.STARTSWITH.value: lambda a, b: a.startswith(b),
+            ACTION.ENDSWITH.value: lambda a, b: a.endswith(b),
+            ACTION.CONTAINS.value: lambda a, b: a in b,
+        }
+
+        try:
+            func = mapping_by_action.get(action, lambda a, b: False)
+            return func(context_value, restriction_value)
+        except Exception:
+            return False
+
+    def _handle_rules(
+        self,
+        logger: object,
+        feature_name: str,
+        rules_context: Dict[str, Any],
+        default_value: bool,
+        rules: List[Dict[str, Any]],
+    ) -> bool:
+        for rule in rules:
+            rule_name = rule.get(RULE_NAME_KEY, "")
+            rule_default_value = rule.get(RULE_DEFAULT_VALUE)
+            is_match = True
+            restrictions: Dict[str, str] = rule.get(RESTRICTIONS_KEY)
+            if restrictions is None:
+                error_str = f"invalid rule schema detected in rule {rule_name}, missing restrictions list"
+                logger.error(error_str)
+                raise ConfigurationException(error_str)
+
+            for restriction in restrictions:
+                context_value = rules_context.get(restriction.get(RESTRICTION_KEY, ""))
+                if not context_value:
+                    logger.debug(
+                        "rule did not not match key",
+                        rule_name=rule_name,
+                        default_value=rule_default_value,
+                        feature_name=feature_name,
+                    )
+                    is_match = False  # rule doesn't match, continue to next one
+                    break
+                elif not self._match_by_action(
+                    restriction.get(RESTRICTION_ACTION), restriction.get(RESTRICTION_VALUE), context_value
+                ):
+                    logger.debug(
+                        "rule did not match action",
+                        rule_name=rule_name,
+                        default_value=rule_default_value,
+                        feature_name=feature_name,
+                    )
+                    is_match = False  # rules doesn't match restriction
+                    break
```

Comment:
it took me a while to think how this loop may work with different inputs. Avoiding two different break points may make the code easier to follow and test. For example, putting the condition of context_value in the _match_by_action function will make this piece easier

Reply:
ok, sounds good.